docs/integrations/tenacity.md

```diff
@@ -0,0 +1,85 @@
+# Tenacity
+
+Making an API call to a provider can fail due to various reasons, such as rate limits, internal server errors, validation errors, and more. Mirascope combined with [Tenacity](https://tenacity.readthedocs.io/en/latest/) increases the chance for these requests to succeed or provides transparency to end users.
+
+## Using Tenacity with Mirascope
+
+A Mirascope call with Tenacity exponential back-off when service is temporarily unavailable:
+
+```python
+from anthropic import InternalServerError
+from tenacity import retry, stop_after_attempt, wait_exponential
+
+from mirascope.core import anthropic, prompt_template
+
+def handle_internal_server_error(errors: list[InternalServerError]) -> None:
+    # Handle internal server errors
+    ...
+
+
+@retry(
+    stop=stop_after_attempt(3),
+    wait=wait_exponential(multiplier=1, min=4, max=10),
+    after=collect_errors(InternalServerError),
+)
+@anthropic.call(model="claude-3-5-sonnet-20240620")
+@prompt_template("What is your purpose?")
+def run(*, errors: list[InternalServerError] = []):
+    handle_internal_server_error(errors)
+
+
+print(run())
+```
+
+Let's break this down. The most common situation is that the call goes through and runs like normal. In the chance that Anthropic is down, we can collect specified `InternalServiceError` using the `collect_errors` function that Mirascope provides. We will make the request again, after waiting for some time and in the meantime call the `handle_internal_server_error` function to inform end users that there's a service outage, or better, make a call to another provider.
```

Comment:
hmm, I'm realizing that the handler here won't actually be called until after the next retry attempt (rather than being called right when it fails), so I'm wondering if (at least until we figure this out) we should just show no collection handling for InternalServiceError and then for validation errors we can show collecting them (and users can extrapolate from there for other errors they may want to include etc.)

Reply:
Is that a bad thing? Seems like its still useful, we could probably use before instead of after right?

Reply:
Yeah it's a bad thing. Let's say we use exponential backoff, and we're waiting 10 seconds for the next call. The handler won't be called until after waiting the 10 seconds at which point we're already making the next call (rather than handling it immediately on failure and then waiting).

I think for now we should just drop this to reduce scope creep and instead just show a basic retry wrap (i.e. no integration) just to show how easy it is to use tenacity with our calls, and then we can show `collect_errors(ValidationError)` for the most common use-case (and currently the only actual known use-case) for collecting errors, which is to reinsert them and the original reason we implemented this.